src/sqlfluff/rules/L039.py

```diff
@@ -34,139 +33,90 @@ class Rule_L039(BaseRule):
-            elif seg.is_type("newline"):
-                prev_newline = True
-                prev_whitespace = None
-            elif seg.is_type("whitespace"):
-                # This is to avoid indents
-                if not prev_newline:
-                    prev_whitespace = seg
-                # We won't set prev_newline to False, just for whitespace
-                # in case there's multiple indents, inserted by other rule
-                # fixes (see #1713)
-            elif seg.is_type("comment"):
-                prev_newline = False
-                prev_whitespace = None
-            else:
-                if prev_whitespace:
-                    if prev_whitespace.raw != " ":
+        # For some segments, strip all whitespace.
+        if context.segment.is_type("object_reference", "comparison_operator"):
+            for child_seg in context.segment.get_raw_segments():
+                if child_seg.is_whitespace:
+                    violations.append(
+                        LintResult(
+                            anchor=child_seg,
+                            fixes=[LintFix.delete(child_seg)],
+                        )
+                    )
+        # Otherwise handle normally
+        else:
+            non_meta_segs = [seg for seg in context.segment.segments if not seg.is_meta]
+            for idx, seg in enumerate(non_meta_segs):
+                if seg.is_type("whitespace"):
+                    # Casting operators shouldn't have any whitespace around them.
+                    # Look for preceeding or following casting operators either as raw
+                    # segments or at the end of a parent segment.
+                    if self._check_diff_idx_for_casting_operator(
+                        non_meta_segs, idx, -1
+                    ) or self._check_diff_idx_for_casting_operator(
+                        non_meta_segs, idx, 1
+                    ):
                         violations.append(
                             LintResult(
-                                anchor=prev_whitespace,
+                                anchor=seg,
                                 fixes=[
-                                    LintFix.replace(
-                                        prev_whitespace,
-                                        [WhitespaceSegment()],
+                                    LintFix.delete(
+                                        seg,
                                     )
                                 ],
                             )
                         )
-                prev_newline = False
-                prev_whitespace = None
-
-            if seg.is_type("object_reference"):
-                for child_seg in seg.get_raw_segments():
-                    if child_seg.is_whitespace:
+                    # If we find whitespace at the start of a segment it's probably
+                    # from a fix, so leave it be. It otherwise shouldn't be there.
+                    elif idx == 0:
```

Comment:
I'm surprised this still happens. A few months back, I modified the `_apply_fixes()` function so any "misplaced" whitespace would be "bubbled up" to an ancestor segment, attempting to force fixes to leave the updated parse tree in a "clean" state. (This was knock-on work from the earlier work to re-parse modified areas of the tree to ensure fixes didn't introduce parse errors.)

Reply:
Yeah - I agree. I'm not going to address this straight away. I'll bring that into the work on reflow. I think that extends the aims of this PR too far.

Reply:
Agreed. I have no boundaries, ha ha. And it's not urgent, more of a 🤔 thing.